main/fancontrol.go

```diff
@@ -0,0 +1,190 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/takama/daemon"
+	"log"
+	"net"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+// #include <wiringPi.h>
+// #cgo LDFLAGS: -lwiringPi
+import "C"
+
+const (
+	// CPU temperature target, degrees C
+	defaultTempTarget = 65.
+	hysteresis        = float32(1.)
+
+	/* This puts our PWM frequency at 19.2 MHz / 2048 =
+	/* 9.4kHz. Higher frequencies will reduce audible switching
+	/* noise but will be less efficient */
+	pwmClockDivisor = 2048
+
+	/* Minimum duty cycle is the point below which the fan does
+	/* not spin. This depends on both your fan and the switching
+	/* transistor used. */
+	defaultPwmDutyMin = 80
+	pwmDutyMax        = 256
+
+	// how often to update
+	delaySeconds = 2
+
+	// GPIO-12 on a Raspberry PI 3
+	defaultPin = 26
+
+	// name of the service
+	name        = "fancontrol"
+	description = "cooling fan speed control based on CPU temperature"
+
+	// port which daemon should be listen
+	port = ":9977"
+)
+
+var stdlog, errlog *log.Logger
+
+func fanControl(pwmDutyMin int, pin int, tempTarget float32) {
+	cPin := C.int(pin)
+	C.wiringPiSetup()
+	C.pwmSetMode(C.PWM_MODE_BAL)
+	C.pinMode(cPin, C.PWM_OUTPUT)
+	C.pwmSetRange(pwmDutyMax)
+	C.pwmSetClock(pwmClockDivisor)
+	C.pwmWrite(cPin, C.int(pwmDutyMin))
+	temp := float32(0.)
+	go cpuTempMonitor(func(cpuTemp float32) {
```

Comment:
IIRC there was one case previously where concurrent access to `/sys/class/thermal/thermal_zone0/temp` caused a kernel panic.